RabbitMQ debug tools

sphuber opened the pull request.

sphuber commented:
Anyone want to review this? @chrisjsewell @ramirezfranciscof @ltalirz @unkcpz ? I would like to get this in 2.1 as it will help users massively with debugging and fixing RabbitMQ issues.

ltalirz commented on aiida/cmdline/commands/cmd_rabbitmq.py:
```diff
@@ -0,0 +1,386 @@
+
+@wrapt.decorator
+@click.pass_context
+def with_client(ctx, wrapped, _, args, kwargs):
```
a significant portion of the code in this file seems quite independent of aiida and could possibly fit inside, say, kiwipy as well.

I'm not saying this has to happen, just worth consideration.
We may also want to comment in the module docstring in why these tools are needed, and under which circumstances we may want to remove them again (if that is the case).

sphuber replied:
Fair, except for the `tasks analysis` command though, that is purely AiiDA specific. But you are right, the rest could be abstracted (currently depends tightly on AiiDA's `Manager` to get the communicator) and be put in `kiwipy`. For me would be fine to move it there at some point if we find it is useful more generically. I had a quick search and there are a few Python libraries out there that provide an interface to the HTTP API, but they are not well maintained.

ltalirz commented on environment.yml:
```diff
@@ -21,14 +21,14 @@ dependencies:
 - kiwipy[rmq]~=0.7.5
 - importlib-metadata~=4.3
 - numpy~=1.19
-- pamqp~=2.3
```
this was not necessary before? or are we removing code that depended on it?

sphuber replied:
We needed it to provide support for RabbitMQ 3.5 but I removed this (see the second commit in this PR). This was its only use so now we no longer need it and can remove it.

ltalirz commented on aiida/cmdline/commands/cmd_rabbitmq.py:
```diff
@@ -0,0 +1,386 @@
+    if not client.is_connected:
+        echo.echo_critical(
+            'Could not connect to the management API. Make sure RabbitMQ is running and the management plugin is '
+            'installed using `sudo rabbitmq-plugins enable rabbitmq_management`.'
```
this worked nicely for me

sphuber replied:
I'm assuming you ran this on a Ubuntu box? Not sure what our mileage will be for other ways of installing RabbitMQ or on other OS's

sphuber replied:
As expected, @unkcpz experienced a problem because he is using RabbitMQ installed on docker. My suspicion is that it doesn't work because the 15672 port which is used by the API is not forwarded. I have updated the error message to include that information.

ltalirz replied:
yes, this was on ubuntu 20.04

ltalirz commented on aiida/cmdline/commands/cmd_rabbitmq.py:
```diff
@@ -0,0 +1,386 @@
+    return pks
+
+
+@cmd_tasks.command('analysis')
```
to stay in line with the names of the other commands, perhaps this should be called `analyze` instead?

sphuber replied:
Yeah, makes sense. Will change it

sphuber replied:
done

ltalirz commented on aiida/manage/external/rmq/client.py:
```diff
@@ -0,0 +1,94 @@
+__all__ = ('RabbitmqManagementClient', 'ManagementApiConnectionError')
+
+
+class ManagementApiConnectionError(AiidaException):
```
also this file looks like it could go to kiwipy

sphuber replied:
Indeed, this is quite a generic API to interface with the management HTTP API. Could abstract this to a standalone library at some point or integrate into `kiwipy`.

ltalirz reviewed (commented):
thanks @sphuber !

I clearly see the usefulness of these tools in helping debug issues, so this is a great addition.

Just some minor questions about where the code should go.
I've tested it locally and it worked fine for me; I have not read through all of the code you added - if there is something in particular you'd like me to look at, let me know.

ltalirz commented:
P.S. I happened to not have any processes running (and the list of rabbitmq tasks is empty), but I still had two queues
```
$ verdi devel rabbitmq queues list
Name                                                    Messages  State
----------------------------------------------------  ----------  -------
aiida-9e50e1e81b3c4dccadabde65a8a745bc.process.queue           0  running
aiida-e86ff3e139f3443a959445e9444f2e3f.process.queue           0  running
```
I guess this points to some issue?

sphuber commented on aiida/cmdline/commands/cmd_rabbitmq.py:
```diff
@@ -0,0 +1,386 @@
+    """Commands related to process tasks."""
+
+
+available_projectors = (
```
good point, there should be. Don't know why `pylint` doesn't complain here. The file is not excluded and the rule is not disabled as far as I can see. Will change it regardless

sphuber replied:
Changed

unkcpz commented:
I am also interested in reviewing this. 

But first I still get a warning with `Warning: RabbitMQ v3.8.14 is not supported and will cause unexpected problems!`. I think `v3.8.14` is supported.
Then I cannot start the management plugin. I using the docker to start the rabbitmq as instructed [here](https://github.com/aiidateam/aiida-core/wiki/RabbitMQ-version-to-use#using-an-old-rabbitmq-version). I start the plugin with the [instruction](https://hub.docker.com/_/rabbitmq), the "Enabling plugins" section but still get the error "Critical: Could not connect to the management API.".

sphuber commented:
> I guess this points to some issue?

No issue here. These are the process task queues and they are declared as being persistent, even if they are empty. Since there is just one per profile, the overhead of keeping them is a lot less then deleting and recreating it each time it is emptied.
 
> Are these scoped by profile?

Yes they are, the UUID you see is the profile UUID.

sphuber commented:
> I am also interested in reviewing this.

That'd be great, thanks!

> But first I still get a warning with `Warning: RabbitMQ v3.8.14 is not supported and will cause unexpected problems!`. I think `v3.8.14` is supported.

I think I added a bug by adding a check on the lower bound, but I inverted the condition. Will fix it now.

> Then I cannot start the management plugin. I using the docker to start the rabbitmq as instructed [here](https://github.com/aiidateam/aiida-core/wiki/RabbitMQ-version-to-use#using-an-old-rabbitmq-version). I start the plugin with the [instruction](https://hub.docker.com/_/rabbitmq), the "Enabling plugins" section but still get the error "Critical: Could not connect to the management API.", any idea how to start it?

The HTTP API is served on port 15672, is that accessible from outside the container? Are you forwarding it just like the 5672 port?

sphuber commented:
> But first I still get a warning with Warning: RabbitMQ v3.8.14 is not supported and will cause unexpected problems!. I think v3.8.14 is supported.

This should now be fixed. I have also added explicit tests for this.

sphuber commented:
@ltalirz thanks for the review. I have addressed your comments. I agree that some of the code is generic enough to be moved to a separate library (`kiwipy` or a standalone one) but don't want to do this now if you don't mind. I don't need you to go through the whole code, but it would be good if you could sign off that you think it is ok to drop support for RabbitMQ 3.5 and older.

unkcpz commented:
> The HTTP API is served on port 15672, is that accessible from outside the container? Are you forwarding it just like the 5672 port?

Ah, true, I forget to forward this port. It works now. Thanks!

ltalirz commented:
> > I guess this points to some issue?
> 
> No issue here. These are the process task queues and they are declared as being persistent, even if they are empty. Since there is just one per profile, the overhead of keeping them is a lot less then deleting and recreating it each time it is emptied.

Would it be possible to add a first column with the profile name that corresponds to the UUID?
Most users probably never see the profile UUID and so they will anyhow need to look it up in order to understand what the output of the command means.

ltalirz commented:
> @ltalirz thanks for the review. I have addressed your comments. I agree that some of the code is generic enough to be moved to a separate library (`kiwipy` or a standalone one) but don't want to do this now if you don't mind. 

Fine

> I don't need you to go through the whole code, but it would be good if you could sign off that you think it is ok to drop support for RabbitMQ 3.5 and older.

Yes. I haven't seen anyone using rabbitmq <3.6 for a very long time and [its EOL was in 2016](https://endoflife.date/rabbitmq).

unkcpz commented on tests/cmdline/commands/test_rabbitmq.py:
```diff
@@ -0,0 +1,138 @@
+    submit_and_await(node)
+    assert node.is_finished_ok
+
+    # Need to manually stop the daemon such that it cleans all state related to the submitted processes. It is not quite
+    # understood how, but leaving the daemon running, a following test that will submit to the daemon may hit an
+    # an exception because the node submitted here would no longer exist and the daemon would try to operate on it.
+    started_daemon_client.stop_daemon(wait=True)
```
If understand correctly, this stops the daemon the second time since `stop_daemon` is already called from `daemon_client` fixture of `started_daemon_client`?

sphuber replied:
Hmm, I think it might indeed no longer be necessary. When I introduced the `daemon_client` fixture, it was originally session-scoped, just for efficiency to only stop it once at the end of the session. But I think exactly for this reason I decided to make it function scoped because it can cause side effects between tests. I can go back and remove it and see if the tests run. They probably will now.

sphuber commented:
> Would it be possible to add a first column with the profile name that corresponds to the UUID?

Possible yes, but not very straightforward. When running normally, these aren't the only queues that there will be, plus RabbitMQ might be used for something other than AiiDA that creates queues that have nothing to do with AiiDA. So for those a profile wouldn't even make sense. I don't think this command will be for normal users anyway. These will at best be for us at developers, so don't think we should tightly couple the output to AiiDA profiles.

unkcpz reviewed (commented):
@sphuber Thanks! This is a very useful feature. I did not review it in detail but just give it a try with some questions I have in mind. 

If I understand correctly, the `devel rabbitmq tasks analyze --fix` will run the `revive_processces` and what the reviev_processes did is to requeue the process?
I run it `analyze --fix` for the lost process (the workchain process that has child process killed by ctrl+C), when I run analyze it again they disappear but still staled process in `verdi process list`, where is the process list read from, do this command need to take care of clean it?

unkcpz commented on aiida/manage/external/rmq/client.py:
```diff
@@ -0,0 +1,94 @@
+        :returns: ``True`` if the server can be reached, ``False`` otherwise.
+        """
+        try:
+            self.request('cluster-name')
```
Where is this `cluster-name` come from, is it hardcode by rabbitmq?

sphuber replied:
This is an endpoint provided by the API. You can see documentation for the API locally at this URL http://localhost:15672/api/index.html

sphuber commented on .github/workflows/ci-code.yml:
```diff
@@ -66,9 +66,10 @@ jobs:
           - 5432:5432
       rabbitmq:
-        image: rabbitmq:latest
+        image: rabbitmq:3.8-management
```
Ah great, will change that then, thanks

sphuber commented:
> If I understand correctly, the `devel rabbitmq tasks analyze --fix` will run the `revive_processces` and what the reviev_processes did is to requeue the process?

Yes, `revive_processes` simply send a new task to the process queue for each of the processes it gets.

> I run it `analyze --fix` for the lost process (the workchain process that has child process killed by ctrl+C), when I run analyze it again they disappear but still staled process in `verdi process list`

Not sure I understand what you mean. The command is not supposed to clean up killed processes. When you say there is still a stale process in `verdi process list` based on what do you say that?

> where is the process list read from, do this command need to take care of clean it?

The `analyze` command gets the active process tasks from the process queue by reading all the tasks it has in it (each task contains the process id it corresponds to). Then it gets the "active" processes from the database, by querying for all process nodes that have a process state in `waiting`, `created` and `running`. Then it does a cross-correlation and check the difference between these sets. Any tasks that do not have a corresponding active node in the database get discarded. Any active process nodes in the database that don't have a corresponding task, get their task recreated ("revived").

unkcpz commented:
> The command is not supposed to clean up killed processes. When you say there is still a stale process in verdi process list based on what do you say that?

It is not the killed processes, they are still all in `Waiting` state and related to the issue https://github.com/aiidateam/aiida-core/issues/3776. I have a workchain run without using the daemon, when I kill it by `ctrl+C` the child process is not properly killed. They are all listed by the `tasks analyze` command, and `--fix` drop them from the list I run `task analyze` second time but they still in the `verdi process list`. But it is the issue of https://github.com/aiidateam/aiida-core/issues/3776 so may not be related to the rabbitmq feature here.

unkcpz reviewed (approved):
It is all good to me. 😄

sphuber commented:
Thanks a lot for the reviews @ltalirz and @unkcpz !
